tool.c

```diff
@@ -221,9 +222,99 @@ compdb(int argc, char *argv[])
 	return 0;
 }
 
+static void
+targets_rule(int argc, char *argv[])
+{
+	struct edge *e;
+	size_t i;
+	for (e = alledges; e; e = e->allnext) {
+		if (argc == 2) {
+			for (i = 0; i < e->nin; ++i)
+				if (!e->in[i]->gen) {
+					printf("%s\n", e->in[i]->path->s);
+			}
+		} else if (argc == 3) {
+			if (strcmp(e->rule->name, argv[2]) == 0) {
+				for (i = 0; i < e->nout; ++i)
+					printf("%s\n", e->out[i]->path->s);
+			}
+		}
+	}
+}
+
+static void
+targets_depth(struct node *n, size_t depth, size_t indent)
+{
+	struct edge *e = n->gen;
+	size_t i;
+
+	for (i = 0; i < indent; ++i)
+		printf("  ");
+	if (e) {
+		printf("%s: %s\n", n->path->s, e->rule->name);
+		if (depth > 1 || depth <= 0) {
+			for (i = 0; i < e->nin; ++i)
+				targets_depth(e->in[i], depth - 1, indent + 1);
+		}
+	} else {
+		printf("%s\n", n->path->s);
+	}
+}
+
+static void
+targets_all(void)
+{
+	struct edge *e = NULL;
+	size_t i;
+	for (e = alledges; e; e = e->allnext) {
+		for (i = 0; i < e->nout; ++i)
+			printf("%s: %s\n", e->out[i]->path->s, e->rule->name);
+	}
+}
+
+static int
+targets(int argc, char *argv[])
+{
+	size_t depth = 1;
+	if (argc >= 2) {
+		const char *mode = argv[1];
+		if (strcmp(mode, "rule") == 0) {
+			targets_rule(argc, argv);
+			return 0;
+		} else if (strcmp(mode, "depth") == 0) {
+			if (argc > 1) {
+				char *end;
+				depth = (size_t)strtol(argv[2], &end, 10);
+				if (*end)
+					fprintf(stderr, "depth not valid");
```

Comment:
This should error out if the depth was invalid.

Also, the format string is missing a trailing newline.